### PR TITLE
LPS-151456 Minimize required FDS configuration

### PR DESCRIPTION
**Pull request by markocikos**
### References

- [LPS-151456 Minimize required FDS configuration](https://issues.liferay.com/browse/LPS-151456)
- [FDS technical improvements checklist](https://liferay.atlassian.net/wiki/spaces/ENGFRONTENDINFRA/pages/2002681985/FDS+technical+improvements+checklist)
- This is a followup on #2114
- This is partially a followup on #2103

### What is the goal of this PR?

FDS tags have unnecessarily required attributes, that we'd like to make optional, along with defaults. The only required attributes of `headless-display` tag will now be `apiURL` and `id`. We plan to do the same for the `classic-display`, in a separate task.

In addition, we are:
- adding a tab in the sample module with minimal configuration FDS. Intent here is to add all the small features to `customized` tab. If there will be a need for a showcase of a larger disrupting feature, we can add it in additional tabs. Note that both FDS instances are using the same data.
- deprecating unused `portletURL` attribute.
- seting portlet namespace as default `namespace`. This makes it [consistent with `<clay>` tags](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/internal/servlet/taglib/BaseContainerTag.java#L142-L158).

:warning: I fully expect this PR to break our CI tests on the sample module, as we are changing the markup. @Esther-OC @mateusralv

### What does it look like?

customized

minimum

### Steps to reproduce

Place sample FDS module on any page.

### CODE REVIEW

**Comment by diegonvs** [comment]
ci:test:relevant

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/BaseDisplayTag.java [hunk 1]
```diff
@@ -129,6 +154,8 @@ protected Map<String, Object> prepareProps(Map<String, Object> props) {
 			props.put("additionalProps", _additionalProps);
 		}
 
+		props.put("namespace", getNamespace());
```
Comment (diegonvs) [inline modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/BaseDisplayTag.java:157]: Hey @markocikos, I don't know what namespace is being used for on FDS but If this should be unique for every FDS displayed on the page. I think we should append a string like UNNAMED_FDS or UNREGISTRED_FDS.

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimum.jsp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+<%@ include file="/init.jsp" %>
+
+<%
+FDSSampleDisplayContext fdsSampleDisplayContext = (FDSSampleDisplayContext)request.getAttribute(FDSSampleWebKeys.FDS_SAMPLE_DISPLAY_CONTEXT);
+%>
+
+<frontend-data-set:headless-display
+	apiURL="<%= fdsSampleDisplayContext.getAPIURL() %>"
+	id="<%= FDSSampleFDSNames.MINIMUM_FDS_SAMPLE %>"
+/>
```
Comment (diegonvs) [inline modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimum.jsp:26]: I liked this idea :)

I think we can emphasize in a text that this page will render a FDS with only required properties

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/MinimumSampleTableFDSView.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	service = FDSView.class
+)
+public class MinimumSampleTableFDSView extends BaseTableFDSView {
+
+	@Override
+	public FDSTableSchema getFDSTableSchema(Locale locale) {
+		FDSTableSchemaBuilder fdsTableSchemaBuilder =
+			_fdsTableSchemaBuilderFactory.create();
+
+		fdsTableSchemaBuilder.addFDSTableSchemaField("id", "id");
```
Comment (beltranrengifo) [inline modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/MinimumSampleTableFDSView.java:45]: Maybe we can also remove some columns in this sample? The `ID` and `Author` for example?

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/SampleCardsFDSView.java [hunk 4]
```diff
@@ -25,7 +25,7 @@
  */
 @Component(
 	enabled = false,
-	property = "frontend.data.set.name=" + FDSSampleFDSNames.FDS_SAMPLES,
+	property = "frontend.data.set.name=" + FDSSampleFDSNames.CUSTOMIZED_FDS_SAMPLE,
 	service = FDSView.class
 )
 public class SampleCardsFDSView extends BaseCardsFDSView {
```
Comment (beltranrengifo) [inline modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/SampleCardsFDSView.java:31]: Does this class name should reference the customized sample, like in `CustomizedSampleTableFDSView`?

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/SampleListFDSView.java [hunk 5]
```diff
@@ -25,7 +25,7 @@
  */
 @Component(
 	enabled = false,
-	property = "frontend.data.set.name=" + FDSSampleFDSNames.FDS_SAMPLES,
+	property = "frontend.data.set.name=" + FDSSampleFDSNames.CUSTOMIZED_FDS_SAMPLE,
 	service = FDSView.class
 )
 public class SampleListFDSView extends BaseListFDSView {
```
Comment (beltranrengifo) [inline modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/SampleListFDSView.java:31]: Does this class name should reference the customized sample, like in CustomizedSampleTableFDSView?

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/SampleTimelineFDSView.java [hunk 6]
```diff
@@ -25,7 +25,7 @@
  */
 @Component(
 	enabled = false,
-	property = "frontend.data.set.name=" + FDSSampleFDSNames.FDS_SAMPLES,
+	property = "frontend.data.set.name=" + FDSSampleFDSNames.CUSTOMIZED_FDS_SAMPLE,
 	service = FDSView.class
 )
 public class SampleTimelineFDSView extends BaseTimelineFDSView {
```
Comment (beltranrengifo) [inline modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/SampleTimelineFDSView.java:31]: Does this class name should reference the customized sample, like in CustomizedSampleTableFDSView?

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/BaseDisplayTag.java (hunk 1 above)
Comment (beltranrengifo) [reply modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/BaseDisplayTag.java:157]: A random string was used in the previous version, not sure if it's the way...

Do we need the `_randomNamespace` variable and related methods now?

**Review by beltranrengifo — COMMENTED** [review]
Hi @markocikos ! 
Thank you for the progress you made here 🙌 
I've left you some minor comments inline

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/MinimumSampleTableFDSView.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * @author Marko Cikos
+ */
+@Component(
+	enabled = true,
+	property = "frontend.data.set.name=" + FDSSampleFDSNames.MINIMUM_FDS_SAMPLE,
+	service = FDSView.class
+)
+public class MinimumSampleTableFDSView extends BaseTableFDSView {
```
[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/MinimumSampleTableFDSView.java:38]: There is currently no way to disable this, we are [unconditionally adding](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/table/TableHead.js#L78-L84) the `FieldsSelectorDropdown` component. Potentially changing this is a product call, see https://liferay.slack.com/archives/C02CH9Q6Q21/p1650366641759489

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/MinimumSampleTableFDSView.java (hunk 3 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/MinimumSampleTableFDSView.java:45]: I think there is some value in keeping the same fields. For example, you can see at glance the difference between default `Author` and `Author` with the custom renderer.

I tried to do is to remove `status` renderer from the status field, to keep consistent default renderer on all fields, but that breaks the component. This is a can of worms we can open in a separate task.

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/SampleCardsFDSView.java (hunk 4 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/SampleCardsFDSView.java:31]: Yeah, that will make it more consistent. I'll also change the constants to match.

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/BaseDisplayTag.java (hunk 1 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/BaseDisplayTag.java:157]: `namespace` does not need to be unique for every FDS. `id` needs to be unique, nothing else. If there is something inside FDS instance that needs to generate a unique value, it should probably be a derivative of `id`.

`namespace` refers to **portlet** namespace. This is consistent over basically all tags in DXP. We can easily have multiple FDS instances in a portlet, as portlets tend to have multiple pages.

Within a DXP tag, namespace is typically required to
1. Build parameters for DXP requests
2. Access DOM elements in a portlets that have been set with `<portlet:namespace />`

Looking at master, `randomNamespace` is not used anywhere :thinking: I think we can safely deprecate it.

Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/BaseDisplayTag.java:157]: On a closer inspection, we are using `randomNamespace` [here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/BaseDisplayTag.java#L128). Admittedly, I don't see a purpose in setting `id` of this `div`, but let's leave this for a separate task.

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimum.jsp (hunk 2 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimum.jsp:26]: I think that tab name is enough. There is only one FDS per tab; if there were several, title per FDS would be useful. A more verbose description in UI might make sense, but that's a slippery slope. If we start with that, we should consistently add description in UI of all tabs. For 'customized' tag, that would have something like 10 lines, cluttering up the page.

If we want a detailed explanation of sample module, what we can do is open a document for this in confluence.

**Comment by markocikos** [comment]
Force-pushed! Changes:
- Renamed model files to match FDS ID. More specifically, if a file contains `property = "frontend.data.set.name=*`, it has a prefix with that name.
- Fixed tabs inclusion logic, so that only one FDS loads with tab.

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
ci:test:relevant

**Review by beltranrengifo — APPROVED** [review]
LGTM!

**Comment by markocikos** [comment]
ci:forward

**Pull request closed without merging**